library/ccm.c

```diff
@@ -269,8 +269,7 @@ int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
                            size_t add_len )
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    unsigned char i;
-    size_t olen, use_len, offset;
+    size_t i, olen, use_len, offset;
```

Comment:
This change is not necessary, the code is perfectly correct as it is

In (new line numbers) line 305, `offset` is set to a value from `0` to `15`

At line 307, `use_len` is set to `16 - offset`, i.e. a value from `1` to `16`

In lines 309 and 310, `use_len` might be reduced still further (if its value is greater than `add_len` - note that `add_len` is a `size_t`, therefore unsigned, therefore we can’t get negative `use_len`

Then in line 312, we have the loop: `for( i = 0; i < use_len; i++ )`

In other words, `i` will have a maximum value of 16 (after the loop is executed), and so `unsigned char` is a perfectly good type for it

If there’s any improvement to be done to this code, it’s reduction of the size and scope of some of these variables, and potentially improving performance with `uint_fastN_t` types, which would look like

```
diff --git a/library/ccm.c b/library/ccm.c
index e06267856..7bbdd0588 100644
--- a/library/ccm.c
+++ b/library/ccm.c
@@ -269,8 +269,6 @@ int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
                            size_t add_len )
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    unsigned char i;
-    size_t olen, use_len, offset;
 
     if( ctx->state & CCM_STATE__ERROR )
     {
@@ -303,14 +301,15 @@ int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
 
         while( add_len > 0 )
         {
-            offset = (ctx->processed + 2) % 16; /* account for y[0] and y[1]
-                                                 * holding total auth data length */
-            use_len = 16 - offset;
+            /* + 2: account for y[0] and y[1] holding total auth data length */
+            uint_fast8_t offset = (uint_fast8_t) ( (ctx->processed + 2) % 16 );
+
+            uint_fast8_t use_len = 16 - offset;
 
             if( use_len > add_len )
                 use_len = add_len;
 
-            for( i = 0; i < use_len; i++ )
+            for( uint_fast8_t i = 0; i < use_len; i++ )
                 ctx->y[i + offset] ^= add[i];
 
             ctx->processed += use_len;
@@ -319,6 +318,7 @@ int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
 
             if( use_len + offset == 16 || ctx->processed == ctx->add_len )
             {
+                size_t olen;
                 if( ( ret = mbedtls_cipher_update( &ctx->cipher_ctx, ctx->y, 16, ctx->y, &olen ) ) != 0 )
                 {
                     ctx->state |= CCM_STATE__ERROR;
```

but I don’t actually think that is worth doing.

Rather than just quoting the Coverity analysis, please inspect the code and see if the Coverity report is a false positive or not.

Reply:
You are right, Coverity got confused by this non-trivial logic that you have detailed, which keeps the iterator limit inside an 8-bit limit. However, the sheer length of your explanation highlights the present code complexity, which I proposed to slightly simplify by keeping all integers the same type, and eliminating such concerns for both human and automated analysis.
Also, I am of the opinion that any 32/64 bit processor offers no performance improvement by reducing scalar variable size below the width of a native register, meaning that neither the original code nor your alternative change proposal really offer a performance advantage.
My solution (as opposed to marking the defect as a false positive) relieves us from the need to justify such false positives to a security auditor, who no doubt would find the analysis that proves this as a false positive somewhat non-trivial.